// @loaders.gl/core/dist/es6/lib/fetch/fetch-file.browser.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import assert from '../../utils/assert';

class FileReadableResponse {
  constructor(fileOrBlob) {
    this._fileOrBlob = fileOrBlob;
    this.bodyUsed = false;
  }

  get headers() {
    return new Headers({
      'Content-Length': this._fileOrBlob.size,
      'Content-Type': this._fileOrBlob.type
    });
  }

  url() {
    return this._fileOrBlob.name || '';
  }

  arrayBuffer() {
    var _this = this;

    return _asyncToGenerator(function* () {
      var {
        reader,
        promise
      } = _this._getFileReader();

      reader.readAsArrayBuffer(_this._fileOrBlob);
      return promise;
    })();
  }

  text() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      var {
        reader,
        promise
      } = _this2._getFileReader();

      reader.readAsText(_this2._fileOrBlob);
      return promise;
    })();
  }

  json() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      var text = yield _this3.text();
      return JSON.parse(text);
    })();
  }

  _getFileReader() {
    assert(!this.bodyUsed);
    this.bodyUsed = true;
    var reader;
    var promise = new Promise((resolve, reject) => {
      try {
        reader = new FileReader();

        reader.onerror = error => reject(new Error(error));

        reader.onabort = () => reject(new Error('Read aborted.'));

        reader.onload = () => resolve(reader.result);
      } catch (error) {
        reject(error);
      }
    });
    return {
      reader,
      promise
    };
  }

}

export default function fetchFileReadable(fileOrBlob, options) {
  return Promise.resolve(new FileReadableResponse(fileOrBlob, options));
}
